### PR TITLE
Fix the bug in gradient clip when the gradient is sparse

### PR DESCRIPTION
**Pull request by chengduoZH**

### CODE REVIEW

File: python/paddle/fluid/clip.py [hunk 1]
```diff
@@ -271,7 +271,7 @@ def _process_context(self, context, param, grad):
                     "All parameters' 'clip_norm' of a same group should be the same"
                 )
 
-        square = grad * grad
+        square = layers.square(grad)
```
Comment (chengduoZH) [inline python/paddle/fluid/clip.py]: In some models, the grad is sparse, but the `sparse * sparse` is meaningless, so I replace with `square`. here.

File: paddle/fluid/operators/activation_op.h [hunk 2]
```diff
@@ -50,16 +56,38 @@ class ActivationKernel
...
+
+    if (CanBeUsedBySelectedRows.count(context.op().Type())) {
+      X = detail::Ref(
+          paddle::framework::GetLoDTensorOrSelectedRowsValueFromVar(*x_var),
+          "Cannot get input Tensor X, variable name = %s",
+          context.op().Input("X"));
+      Out = paddle::framework::GetMutableLoDTensorOrSelectedRowsValueFromVar(
+          out_var);
+    } else {
+      X = detail::Ref(context.Input<framework::Tensor>("X"),
```
Comment (panyx0718) [inline paddle/fluid/operators/activation_op.h:78]: Does this throw error if X is selected rows (Cannot be used by SelectedRows.

**Review by panyx0718 — COMMENTED** [review]
API is fine. Some comments

File: python/paddle/fluid/clip.py [hunk 3]
```diff
@@ -271,7 +271,12 @@ def _process_context(self, context, param, grad):
...
                 )
 
-        square = grad * grad
+        merge_grad = grad
+        if grad.type == core.VarDesc.VarType.SELECTED_ROWS:
+            merge_grad = layers.merge_selected_rows(grad)
+            merge_grad = layers.get_tensor_from_selected_rows(merge_grad)
+
+        square = layers.square(merge_grad)
         local_norm_var = layers.reduce_sum(input=square)
```
Comment (chengduoZH) [inline python/paddle/fluid/clip.py:280]: `reduce_sum ` only accepts `LoDTensor`, so, if the grad is SELECTED_ROWS, there should use `get_tensor_from_selected_rows`.

File: paddle/fluid/operators/activation_op.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/operators/activation_op.h]: yup

Comment (chengduoZH) [reply paddle/fluid/operators/activation_op.h:78]: Yup

**Review by panyx0718 — APPROVED** [review]

File: paddle/fluid/operators/merge_selected_rows_op.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+             "The input type is SelectedRows, and the selected rows may be "
+             "duplicated.");
+    AddOutput("Out",
+              "The output type is SelectedRows, and the selected rows are not "
+              "duplicated.");
+    AddComment(
+        R"DOC(
+MergeSelectedRows Operator.
+
+MergeSelectedRows is used to merge the duplicated rows of the input.
```
Comment (qingqing01) [inline paddle/fluid/operators/merge_selected_rows_op.cc:46]: Better to describe clearly how to merge. (add duplicated rows ?)

**Review by qingqing01 — COMMENTED** [review]
Only reviewed the API.  LGTM.

File: paddle/fluid/operators/merge_selected_rows_op.cc (hunk 4 above)
Comment (chengduoZH) [reply paddle/fluid/operators/merge_selected_rows_op.cc:46]: Thanks, I will add more clear description in next PR.

**Review by typhoonzero — APPROVED** [review]
Sometimes people may think "merge" should have two inputs instead of one, it's actually doing dedup and sum. Since the internal implement is called "MergeAdd" I think it's fine to keep the name.

**Comment by chengduoZH** [comment]
@typhoonzero Thanks, I will fix in next PR

**Pull request merged**